core/src/main/java/org/elasticsearch/cluster/routing/RoutingChangesObserver.java

```diff
@@ -0,0 +1,196 @@
+
+    /**
+     * Called when a shard is failed or cancelled.
+     */
+    void shardFailed(ShardRouting failedShard, UnassignedInfo unassignedInfo);
+
+    /**
+     * Called on relocation source when relocation completes after relocation target is started.
+     */
+    void relocationCompleted(ShardRouting removedRelocationSource);
+
+    /**
+     * Called on replica relocation target when replica relocation source fails. Promotes the replica relocation target to ordinary
+     * initializing shard.
+     */
+    void relocationSourceRemoved(ShardRouting removedReplicaRelocationSource);
+
+    /**
+     * Called on started primary shard after it has been promoted from replica to primary and is reinitialized due to shadow replicas.
+     */
+    void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardRouting initializedShard);
+
+    /**
+     * Called when started replica is promoted to primary.
+     */
+    void replicaPromoted(ShardRouting replicaShard);
+
+
+    /**
+     * Abstract implementation of {@link RoutingChangesObserver} that does not take any action. Useful for subclasses that only override
+     * certain methods.
+     */
+    class AbstractRoutingChangesObserver implements RoutingChangesObserver {
+
+        @Override
+        public void shardInitialized(ShardRouting unassignedShard) {
+
+        }
+
+        @Override
+        public void shardStarted(ShardRouting initializingShard, ShardRouting startedShard) {
+
+        }
+
+        @Override
+        public void relocationStarted(ShardRouting startedShard, ShardRouting targetRelocatingShard) {
+
+        }
+
+        @Override
+        public void unassignedInfoUpdated(ShardRouting unassignedShard, UnassignedInfo newUnassignedInfo) {
+
+        }
+
+        @Override
+        public void shardFailed(ShardRouting activeShard, UnassignedInfo unassignedInfo) {
+
+        }
+
+        @Override
+        public void relocationCompleted(ShardRouting removedRelocationSource) {
+
+        }
+
+        @Override
+        public void relocationSourceRemoved(ShardRouting removedReplicaRelocationSource) {
+
+        }
+
+        @Override
+        public void startedPrimaryReinitialized(ShardRouting startedPrimaryShard, ShardRouting initializedShard) {
+
+        }
+
+        @Override
+        public void replicaPromoted(ShardRouting replicaShard) {
+
+        }
+    }
+
+    class DelegatingRoutingChangesObserver implements RoutingChangesObserver {
+
+        private final RoutingChangesObserver[] routingChangesObservers;
+
+        public DelegatingRoutingChangesObserver(RoutingChangesObserver... routingChangesObservers) {
+            this.routingChangesObservers = routingChangesObservers;
+        }
+
+        @Override
+        public void shardInitialized(ShardRouting unassignedShard) {
+            for (RoutingChangesObserver routingChangesObserver : routingChangesObservers) {
+                routingChangesObserver.shardInitialized(unassignedShard);
```

Comment:
So for the observers, should we continue to notify other observers if one fails? Should we catch exceptions from each observer and re-throw at the end? I think if we decide one exceptional observer halts further observations we should document that somewhere - we don't want a failure in the snapshot/restore observer to cause a problem with allocation due to a misplaced exception